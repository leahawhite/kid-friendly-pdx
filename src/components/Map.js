import React from 'react';
import ReactDOM from 'react-dom';
import { camelize } from './helpers'

export class Map extends React.Component {
  constructor(props) {
    super(props);

    const {lat, lng} = this.props.initialCenter;
    this.state = {
      currentLocation: {
        lat: lat,
        lng: lng
      }
    }
  }
  

  componentDidMount() {
    if (this.props.centerAroundCurrentLocation) {
      if (navigator && navigator.geolocation) {
        navigator.geolocation.getCurrentPosition((pos) => {
          const coords = pos.coords;
          this.setState({
            currentLocation: {
              lat: coords.latitude,
              lng: coords.longitude
            }
          })
        })
      }
    }
    this.loadMap();
  }
  
  componentDidUpdate(prevProps, prevState) {
    if (prevProps.google !== this.props.google) {
      this.loadMap();
    }
    if (prevState.currentLocation !== this.state.currentLocation) {
      this.recenterMap();
    }
  }

  loadMap() {
    if (this.props && this.props.google) {
      // google is available
      const {google} = this.props;
      const maps = google.maps;

      const mapRef = this.refs.map;
      const node = ReactDOM.findDOMNode(mapRef);

      let {initialCenter, zoom} = this.props;
      const {lat, lng} = this.state.currentLocation;
      const center = new maps.LatLng(lat, lng);
      const mapConfig = Object.assign({}, {
        center: center,
        zoom: zoom
      })
      this.map = new maps.Map(node, mapConfig);
    
      const evtNames = ['ready', 'click', 'dragend'];
      evtNames.forEach(e => {
        this.map.addListener(e, this.handleEvent(e));
      });

      maps.event.trigger(this.map, 'ready');
      let centerChangedTimeout;
  }
  

  handleEvent(evtName) {
    let timeout;
    const handlerName = `on${camelize(evtName)}`;

    return (e) => {
      if (timeout) {
        clearTimeout(timeout);
        timeout = null;
      }
      timeout = setTimeout(() => {
        if (this.props[handlerName]) {
          this.props[handlerName](this.props, this.map, e);
        }
      }, 0);
    }
  }

  recenterMap() {
    const map = this.map;
    const curr = this.state.currentLocation;

    const google = this.props.google;
    const maps = google.maps;

    if (map) {
      let center = new maps.LatLng(curr.lat, curr.lng);
      map.panTo(center);
    }
  }

  renderChildren() {
    const {children} = this.props;

    if (!children) return;

    return React.Children.map(children, c => {
      return React.cloneElement(c, {
        map: this.map,
        google: this.props.google,
        mapCenter: this.state.currentLocation
      });
    })
  }

  render() {
    return (
      <div ref='map'>
        Loading map...
        {this.renderChildren()}
      </div>
    )
  }
}

Map.propTypes = {
  google: React.PropTypes.object,
  zoom: React.PropTypes.number,
  initialCenter: React.PropTypes.object,
  centerAroundCurrentLocation: React.PropTypes.bool,
  onMove: React.PropTypes.func,
  // evtNames.forEach(e => Map.propTypes[camelize(e)] = T.func)
}

Map.defaultProps = {
  zoom: 13,
  // San Francisco, by default
  initialCenter: {
    lat: 37.774929,
    lng: -122.419416
  },
  centerAroundCurrentLocation: false,
  onMove: function() {} // default prop
  
}