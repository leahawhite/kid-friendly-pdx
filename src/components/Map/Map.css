.infowindow {
  font-size: 14px;
  text-align: left;
  overflow: hidden;
  /* width: 200px; */
}

.infowindow-hidden {
  display: none;
}

.infowindow-content {
  text-align: left;
}

.infowindow-place-name {
  padding: 0;
  margin: 0;
}

.infowindow-name-sm {
  font-size: 12px;
}

.infowindow-image-container {
  height: 50px;
  width: 50px; 
  float: right;
}

.infowindow-place-image {
  height: 100%;
}

.infowindow-address {
  font-size: 14px;
}

