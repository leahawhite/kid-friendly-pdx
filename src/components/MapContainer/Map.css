.infowindow {
  font-size: 14px;
  text-align: left;
  overflow: hidden;
  width: 250px;
}

.infowindow-content {
  text-align: left;
}

.infowindow place-name {
  font-size: 16px;
}

.infowindow-image-container {
  height: 50px;
  width: 50px; 
  float: right;
}

.infowindow-place-image {
  height: 100%;
}

.infowindow-address {
  font-size: 14px;
}