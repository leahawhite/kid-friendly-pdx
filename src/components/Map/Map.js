import React, { Component } from 'react'
import { compose } from "recompose"
import { GoogleMap, withScriptjs, withGoogleMap, Marker, InfoWindow } from 'react-google-maps'
import { StarRating } from '../StarRating/StarRating';
import { readableReviewCount } from '../../helpers/helpers';
import config from '../../config';
import './Map.css'

const MapWithMarkers = compose(withScriptjs, withGoogleMap, 
  /*lifecycle({
  componentWillMount() {
    this.setState({
      zoomToMarkers: map => {
        const bounds = new window.google.maps.LatLngBounds();
        map.props.children.forEach((child) => {
          if (child.type === Marker) {
            bounds.extend(new window.google.maps.LatLng(child.props.position.lat, child.props.position.lng));
          }
        })
        map.fitBounds(bounds);
      }
    })
  },
}),*/)(props => {
  
  return (
    <GoogleMap ref={props.zoomToMarkers} defaultZoom={props.zoom} center={props.center} >
      {props.places.map(place => {
        const onMouseOver = props.onMouseOver.bind(this, place)
        const onMouseOut = props.onMouseOut.bind(this, place)
        const onClick = props.onClick.bind(this, place)
        const onClickClose = props.onClickClose.bind(this, place)
        
        return (
          <Marker
            key={place.id}
            position={{
              lat: place.coordinates.lat,
              lng: place.coordinates.lng
            }}
            onMouseOver={onMouseOver}
            onMouseOut={onMouseOut}
            onClick={onClick}
            onClickClose={onClickClose}
            infoClass={props.infoClass}
          >
            {props.selectedPlace === place &&
              <InfoWindow
                onMouseOut={onMouseOut}
                onClickClose={onClickClose}
                infoClass={props.infoClass}
              >
                <div className="infowindow">
                  <h3 className="infowindow-place-name">{place.name}</h3>
                  <div className={props.infoClass}>
                    <div className="infowindow-image-container">
                      {place.images.length ? <img className="infowindow-place-image" src={place.images[0].src} alt={place.images[0].title}></img> 
                      : null}
                    </div>
                    <div className="place-info-reviews">
                      <div className="star-rating">
                        <StarRating rating={place.average_review_rating} />
                      </div>
                      <span>{readableReviewCount(place.number_of_reviews)}</span>
                      <p className="infowindow-address">
                        {place.address1}{place.address2}<br/>{place.city}{', '}{place.state}{' '}{place.zipcode}</p>
                    </div>
                  </div>
                </div>
              </InfoWindow>
            }
          </Marker>
        )
      })}
    </GoogleMap>
  )
})

export default class Map extends Component {
  constructor(props) {
    super(props)
    this.state = {
      selectedPlace: null,
    }
  }

  handleMouseOver = (place, event) => {
    this.setState({ selectedPlace: place })
  }

  handleMouseOut = (place, event) => {
    this.setState({ selectedPlace: null })
  }

  handleClick = (place, event) => {
    this.setState({ selectedPlace: place })
  }

  handleClickClose = (place, event) => {
    this.setState({ selectedPlace: null })
  }

  render() {
    return (
      <MapWithMarkers
        selectedPlace={this.state.selectedPlace}
        places={this.props.places}
        zoom={this.props.zoom}
        center={this.props.center}
        onClick={this.handleClick}
        onClickClose={this.handleClickClose}
        onMouseOver={this.handleMouseOver}
        onMouseOut={this.handleMouseOut}
        infoClass={this.props.infoClass}
        googleMapURL={`https://maps.googleapis.com/maps/api/js?key=${config.GAPI_KEY}&libraries=places`}
        loadingElement={<div style={{ height: `100%` }} />}
        containerElement={<div style={{ height: `250px` }} />}
        mapElement={<div style={{ height: `100%` }} />}
      />
    )
  }
}