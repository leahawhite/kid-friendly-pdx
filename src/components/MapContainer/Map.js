import React, { Component } from 'react'
import { compose } from "recompose"
import { Link } from 'react-router-dom'
import { GoogleMap, withScriptjs, withGoogleMap, Marker, InfoWindow } from 'react-google-maps'
import { StarRating } from '../StarRating/StarRating';
import { readableReviewCount } from '../../helpers/helpers';
import config from '../../config';
import './Map.css'

const MapWithAMarker = compose(withScriptjs, withGoogleMap)(props => {

  return (
    <GoogleMap defaultZoom={12} center={{ lat: 45.5155, lng: -122.6793 }}>
      {props.places.map(place => {
        const onMouseOver = props.onMouseOver.bind(this, place)
        const onMouseOut = props.onMouseOut.bind(this, place)
        return (
          <Marker
            key={place.id}
            position={{
              lat: place.coordinates.lat,
              lng: place.coordinates.lng
            }}
            onMouseOver={onMouseOver}
            // onMouseOut={onMouseOut}
          >
            {props.selectedPlace === place &&
              <InfoWindow
                onMouseOut={onMouseOut}
                maxWidth={0}
              >
                <div className="infowindow">
                  <h3 className="infowindow place-name">
                    <Link to={{
                      pathname: `/places/${place.id}`,
                      state: { place: place }
                      }}>
                    {place.name}
                    </Link>
                  </h3>
                  <div className="infowindow-image-container">
                    {place.images.length ? <img className="infowindow-place-image" src={place.images[0].src} alt={place.images[0].title}></img> 
                    : null}
                  </div>
                  <div className="place-info-reviews">
                    <div className="star-rating">
                      <StarRating rating={place.average_review_rating} />
                    </div>
                    <span>{readableReviewCount(place.number_of_reviews)}</span>
                    <p className="infowindow-address">
                      {place.address1}{place.address2}<br/>{place.city}{', '}{place.state}{' '}{place.zipcode}</p>
                  </div>
                </div>
              </InfoWindow>
            }
          </Marker>
        )
      })}
    </GoogleMap>
  )
})

export default class Map extends Component {
  constructor(props) {
    super(props)
    this.state = {
      places: this.props.places,
      selectedPlace: null
    }
  }

  handleMouseOver = (place, event) => {
    this.setState({ selectedPlace: place })
  }

  handleMouseOut = (place, event) => {
    this.setState({ selectedPlace: null })
  }

  render() {
    return (
      <MapWithAMarker
        selectedPlace={this.state.selectedPlace}
        places={this.state.places}
        onMouseOver={this.handleMouseOver}
        onMouseOut={this.handleMouseOut}
        googleMapURL={`https://maps.googleapis.com/maps/api/js?key=${config.GAPI_KEY}&libraries=places`}
        loadingElement={<div style={{ height: `100%` }} />}
        containerElement={<div style={{ height: `400px` }} />}
        mapElement={<div style={{ height: `100%` }} />}
      />
    )
  }
}