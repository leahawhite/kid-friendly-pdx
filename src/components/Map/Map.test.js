import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Map from './Map';

describe('Map component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Map />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders UI as expected', () => {
    const wrapper = shallow(<Map />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})