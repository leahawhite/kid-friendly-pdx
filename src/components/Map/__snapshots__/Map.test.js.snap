// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map component renders UI as expected 1`] = `
<withScriptjs(withGoogleMap(Component))
  containerElement={
    <div
      style={
        Object {
          "height": "250px",
        }
      }
    />
  }
  googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyDsdDZ5P1zY_soUX1qBLCTrVuN7HENqhjs&libraries=places"
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  mapElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  onClick={[Function]}
  onClickClose={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  selectedPlace={null}
/>
`;
